src/handle.rs

```diff
@@ -70,155 +84,177 @@ impl Handle {
-        let mut bpf = bpf;
-
-        let ret_code = unsafe { pcap_sys::pcap_setfilter(self.handle, bpf.inner_mut()) };
-        if ret_code != 0 {
-            return Err(pcap_util::convert_libpcap_error(self.handle));
+    pub fn get_datalink(&self) -> Result<i32, Error> {
+        let r = unsafe { pcap_sys::pcap_datalink(self.handle) };
+        if r < 0 {
+            Err(pcap_util::convert_libpcap_error(self.handle))
+        } else {
+            Ok(r)
         }
-        Ok(self)
     }
 
-    pub fn set_immediate_mode(&self) -> Result<&Self, Error> {
+    pub fn set_immediate_mode(self) -> Result<Self, Error> {
         if 0 != unsafe { pcap_sys::pcap_set_immediate_mode(self.handle, 1) } {
             Err(pcap_util::convert_libpcap_error(self.handle))
         } else {
             Ok(self)
         }
     }
 
-    pub fn activate(&self) -> Result<&Self, Error> {
-        if 0 != unsafe { pcap_sys::pcap_activate(self.handle) } {
-            Err(pcap_util::convert_libpcap_error(self.handle))
-        } else {
-            Ok(self)
+    pub fn activate(self) -> Result<Handle, Error> {
+        let h = Handle {
+            handle: self.handle,
+            live_capture: self.live_capture,
+            interrupted: std::sync::Arc::new(std::sync::Mutex::new(false)),
+        };
+        if self.live_capture {
+            if 0 != unsafe { pcap_sys::pcap_activate(h.handle) } {
+                return Err(pcap_util::convert_libpcap_error(h.handle));
+            }
         }
+        Ok(h)
     }
+}
 
-    pub fn fd(&self) -> Result<i32, Error> {
-        unsafe {
-            let fd = pcap_sys::pcap_get_selectable_fd(self.handle);
-            if fd == -1 {
-                Err(pcap_util::convert_libpcap_error(self.handle))
-            } else {
-                Ok(fd)
-            }
+impl std::convert::TryFrom<&Config> for PendingHandle {
+    type Error = Error;
+
+    fn try_from(v: &Config) -> Result<Self, Self::Error> {
+        let mut pending = match v.interface() {
+            Interface::Dead { linktype, snaplen } => PendingHandle::dead(*linktype, *snaplen)?,
+            Interface::Lookup => PendingHandle::lookup()?,
+            Interface::File(path) => PendingHandle::file_capture(path)?,
+            Interface::Live(dev) => PendingHandle::live_capture(dev)?,
+        };
+
+        if pending.live_capture {
+            pending = pending
+                .set_snaplen(v.snaplen())?
+                .set_promiscuous()?
+                .set_buffer_size(v.buffer_size())?;
         }
+
+        Ok(pending)
     }
+}
 
-    pub fn as_mut_ptr(&self) -> *mut pcap_sys::pcap_t {
-        self.handle
+/// Wrapper around a pcap_t handle to indicate live or offline capture, and allow the handle to
+/// be interrupted to stop capture.
+#[derive(Clone)]
+pub struct Handle {
+    handle: *mut pcap_sys::pcap_t,
+    live_capture: bool,
+    interrupted: std::sync::Arc<std::sync::Mutex<bool>>,
```

Comment:
Have you considered using a `Arc<AtomicBool>` here?

Reply:
Changed.